src/engine/MultiColumnJoin.cpp

```diff
@@ -0,0 +1,471 @@
+// Copyright 2018, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Florian Kramer (florian.kramer@netpun.uni-freiburg.de)
+
+#include "./MultiColumnJoin.h"
```

Comment:
Does it make sense to have this separate from the normal join? Do you think that the performance penalty of using this more general code for one column over the specialized one-column code is worth the extra maintenance burden?

Reply:
Looking especially [here](https://github.com/ad-freiburg/QLever/blob/63a4212c4f347f7690d7c90731e13ffd926ccf9b/src/engine/Join.cpp#L52) that code looks quite difficult to maintain. I mean the `MultiColumnJoinCaller` magic isn't entirely easy to understand, but at least it should be harder to get only some cases wrong.

That said replacing the normal `Join` should probably be medium term so maybe we just want to add a `TODO` to replace the other `Join` when we are reasonably confident in this one.

Reply:
I added TODOs into the QueryPlanner where the joins are being created. Given how essential joins are in the query evaluation and how stable the code has been I'd say we should be careful when replacing the join operation.
We should definitely do some benchmarks on the join performance. Currently the normal join has the advantage of using a galloping join for columns of hugely different sizes, so before we benchmark the joins we would also need to implement that for the MultiColumnJoin.